airbyte-integrations/connectors/source-mailgun/source_mailgun/source.py

```diff
@@ -0,0 +1,178 @@
+        default_shift = 60 * 60 * 24 * 3
+        self.start_timestamp = config.get('timestamp', time.time() - default_shift)
+        self.ascending = True
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> \
+            Mapping[str, Any]:
+        try:
+            current_parsed_timestamp = current_stream_state['timestamp']
+            latest_record_timestamp = latest_record['timestamp']
+            return {'timestamp': max(current_parsed_timestamp, latest_record_timestamp)}
+        except (KeyError, TypeError):
+            return {'timestamp': self.start_timestamp}
+
+    def path(self, *args, next_page_token: Optional[Mapping[str, Any]] = None, **kwargs) -> str:
+        # return super().path(*args, **kwargs) or "events"  # TODO: Requires the CDK update (see MailgunStream.path())
+        path: str = "events"
+        if next_page_token:
+            token: str = next_page_token['url'].rpartition('/')[2]
+            path = f"{path}/{token}"
+
+        return path
+
+    def request_params(
+            self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None,
+            next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        params = super().request_params(
+            stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token
+        )
+        stream_slice = stream_slice or {
+            "begin": stream_state["timestamp"],
+            "ascending": "yes" if self.ascending else "no",
+        }
+        params.update(stream_slice)
+        return params
+
+    def stream_slices(self, stream_state: Mapping[str, Any] = None, **kwargs) -> Iterable[Optional[Mapping[str, float]]]:
+        """
+        Provide a generator of date_slices in such format:
+        {"begin": 1636411500.0, "end": 1636497900.0}
+        """
+        stream_state = stream_state or {}
+        start_date = stream_state.get(self.cursor_field, self.start_timestamp)
+        for period in self.chunk_timestamps_range(start_date):
+            yield {
+                "begin": period[0], "end": period[1]
+            }
+
+
+# Source
+class SourceMailgun(AbstractSource):
+    def check_connection(self, logger, config) -> Tuple[bool, any]:
+        """
+        :param config:  the user-input config object conforming to the connector's spec.json
+        :param logger:  logger object
+        :return Tuple[bool, any]: (True, None) if the input config can be used to connect to the API successfully, (False, error) otherwise.
+        """
+        try:
+            response = requests.get(urljoin(MailgunStream.url_base, 'domains'), auth=('api', config['private_key']))
+            if response.status_code == 200:
+                return True, None
+            else:
+                message = "Check connection is failed. "
```

Comment:
this will be self-repeating, we know that the check connection is failed because it happens in the check connection, also user will know this because it happened when he click `check`.
but ok, this might be arguable, see my next comment below